addons/sale_timesheet/models/sale_order.py

```diff
@@ -42,14 +42,13 @@ def _compute_timesheet_ids(self):
 
     @api.depends('timesheet_ids', 'company_id.timesheet_encode_uom_id')
     def _compute_timesheet_total_duration(self):
+        group_data = self.env['account.analytic.line'].sudo().read_group([
+            ('order_id', 'in', self.ids)
+        ], ['order_id', 'unit_amount'], ['order_id'])
+        timesheet_unit_amount_dict = defaultdict(float)
+        timesheet_unit_amount_dict.update({data['order_id'][0]: data['unit_amount'] for data in group_data})
         for sale_order in self:
-            timesheets = sale_order.timesheet_ids if self.user_has_groups('hr_timesheet.group_hr_timesheet_approver') else sale_order.timesheet_ids.filtered(lambda t: t.user_id.id == self.env.uid)
-            total_time = 0.0
-            for timesheet in timesheets:
-                # Timesheets may be stored in a different unit of measure, so first we convert all of them to the reference unit
-                total_time += timesheet.unit_amount * timesheet.product_uom_id.factor_inv
-            # Now convert to the proper unit of measure
-            total_time *= sale_order.timesheet_encode_uom_id.factor
+            total_time = sale_order.company_id.project_time_mode_id._compute_quantity(timesheet_unit_amount_dict[sale_order.id], sale_order.timesheet_encode_uom_id)
```

Comment:
Great !

@IT-Ideas you may be interested inhere, I let a comment just to have your attention :)

Reply:
It is normal we no longer check the access right of the current user ? 🤔

Reply:
@xavierbol checking access right cause this problem [Task](https://www.odoo.com/web#id=2514517&action=4043&model=project.task&view_type=form&cids=2&menu_id=4720).
user will only see a stat button when he has access right for a timesheet app
and about hours that show on that stat button should be total hours that deliver for that sale order if users have the least access right for timesheet app.

Reply:
1. *user will only see a stat button when he has access right for a timesheet app*
Ok, we could not compute the total_time if we are sure the user has not the access right. But if we do this, then we need to be sure we don't use `timesheet_total_duration` field in the code.
If we choose to compute it even the user has not the access to see the access right to see the stat button then it is ok for this point.

2. *and about hours that show on that stat button should be total hours that deliver for that sale order if users have the least access right for timesheet app.*
I don't really agree with you, if the user can only see his timesheets then we will show for instance 15 hours recorded in the stat button and when the user clicks on this button, he will only see 5 hours recorded in the view list of timesheets because he can see only his timesheets.
I asked with AJU and she agree with me.

Reply:
@xavierbol 
1. I agree with you. timesheet_total_duration field is not used anywhere so we can not compute the totla_time when user has no access right for a timesheet app.
2. I didn't get you. can you, please elaborate?.

Reply:
@mpa-odoo If the user has only the `group_hr_timesheet_user` group, then he can only see his own timesheets. When this user goes to the form view containing this stat button, he will see the total hours recorded for **all timesheets** for this record (in this case, it is a SO) in the form view. So, when he will click on this stat button, we will filter the timesheets recorded for this SO to only display the ones he can see. For instance, if there are 2 timesheets done by this user and 3 other ones for another user. The total duration displayed in the stat button will not be the total hours recorded in the list view of timesheets displayed when he clicks on this stat button.

I hope it is clear for you.